libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1018 @@
+      {}
+
+      ~iterator() {
+         if (itr) {
+            internal_use_do_not_use::kv_it_destroy(itr);
+         }
+      }
+
+      iterator& operator=(iterator&& other) {
+         itr = std::exchange(other.itr, itr);
+         itr_stat = std::move(other.itr_stat);
+         config = std::move(other.config);
+
+         return *this;
+      }
+
+      /**
+       * Returns the value that the iterator points to.
+       * @ingroup keyvalue
+       *
+       * @return The value that the iterator points to.
+       */
+      T value() const {
+         using namespace detail;
+
+         eosio::check(itr_stat != kv_it_stat::iterator_end, "Cannot read end iterator");
+
+         uint32_t value_size;
+         uint32_t actual_value_size;
+         uint32_t offset = 0;
+
+         // call once to get the value_size
+         internal_use_do_not_use::kv_it_value(itr, 0, (char*)nullptr, 0, value_size);
+
+         void* buffer = value_size > detail::max_stack_buffer_size ? malloc(value_size) : alloca(value_size);
+         auto stat = internal_use_do_not_use::kv_it_value(itr, offset, (char*)buffer, value_size, actual_value_size);
+
+         eosio::check(static_cast<kv_it_stat>(stat) == kv_it_stat::iterator_ok, "Error reading value");
+
+         void* deserialize_buffer = buffer;
+         size_t deserialize_size = actual_value_size;
+
+         if (config.index_name != config.primary_index_name) {
+            uint32_t actual_data_size;
+            auto success = internal_use_do_not_use::kv_get(config.db_name, config.contract_name.value, (char*)buffer, actual_value_size, actual_data_size);
+            eosio::check(success, "failure getting primary key");
+
+            void* pk_buffer = actual_data_size > detail::max_stack_buffer_size ? malloc(actual_data_size) : alloca(actual_data_size);
+            internal_use_do_not_use::kv_get_data(config.db_name, 0, (char*)pk_buffer, actual_data_size);
+
+            deserialize_buffer = pk_buffer;
+            deserialize_size = actual_data_size;
+         }
+
+         T val;
+         deserialize(val, deserialize_buffer, deserialize_size);
+         return val;
+      }
+
+      iterator& operator++() {
+         eosio::check(itr_stat != kv_it_stat::iterator_end, "cannot increment end iterator");
+         itr_stat = static_cast<kv_it_stat>(internal_use_do_not_use::kv_it_next(itr));
+         return *this;
+      }
+
+      iterator& operator--() {
+         if (!itr) {
+            itr = internal_use_do_not_use::kv_it_create(config.db_name, config.contract_name.value, config.prefix.data(), config.prefix.size());
+         }
+         itr_stat = static_cast<kv_it_stat>(internal_use_do_not_use::kv_it_prev(itr));
+         eosio::check(itr_stat != kv_it_stat::iterator_end, "decremented past the beginning");
+         return *this;
+      }
+
+      uint32_t key_compare(key_type kt) const {
+         return internal_use_do_not_use::kv_it_key_compare(itr, kt.data(), kt.size());
```

Comment:
`kv_it_compare` returns int32_t.

Reply:
Also, this behaves differently when ``itr == 0`` from when ``itr`` is an end iterator.

Reply:
In the case where `itr==0` and it's not an end iterator, it should be fine just to return some negative number to signal `less than` right?

Reply:
> itr==0 and it's not an end iterator

Impossible.  ``itr == 0`` is an end iterator, from the wrapper's perspective.

Reply:
Then I'm not sure I understand what you mean by
> Also, this behaves differently when itr == 0 from when itr is an end iterator.